File: android/app/src/main/java/com/zillit/drive/util/crypto/BodyHashGenerator.kt
package com.zillit.drive.util.crypto

import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import java.security.MessageDigest

/**
 * Generates the bodyhash header matching the web app's generateBodyHash function.
 *
 * Web implementation (multipleFunction.js:1808-1818):
 *   1. Build object: { payload: requestBody, moduledata: encryptedModuleData }
 *   2. JSON.stringify that object
 *   3. Append the IV string (salt)
 *   4. SHA-256 hash the combined string
 *   5. Hex-encode the hash
 */
object BodyHashGenerator {

    private val moshi = Moshi.Builder().build()
    private val mapAdapter = moshi.adapter<Map<String, Any?>>(
        Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java)
    )

    fun generate(requestBody: Any?, moduledata: String, salt: String): String {
        // Build the moduleDataString object matching JS: { payload: requestBody, moduledata }
        val moduleDataMap = mapOf(
            "payload" to (requestBody ?: ""),
            "moduledata" to moduledata
        )

        // JSON.stringify equivalent
        val stringToHash = mapAdapter.toJson(moduleDataMap)

        // Append salt (IV string)
        val combinedString = stringToHash + salt

        // SHA-256 hash and hex-encode
        return hashWithSha256(combinedString)
    }

    private fun hashWithSha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(input.toByteArray(Charsets.UTF_8))
        return hashBytes.joinToString("") { "%02x".format(it) }
    }
}

File: ios/ZillitDrive/Utilities/Crypto/BodyHashGenerator.swift
import Foundation
import CryptoKit

/// Generates the bodyhash header matching the web app's generateBodyHash function.
///
/// Web implementation (multipleFunction.js:1808-1818):
///   1. Build object: { payload: requestBody, moduledata: encryptedModuleData }
///   2. JSON.stringify that object
///   3. Append the IV string (salt)
///   4. SHA-256 hash the combined string
///   5. Hex-encode the hash
enum BodyHashGenerator {

    static func generate(requestBody: Any?, moduledata: String, salt: String) -> String {
        // Build the moduleDataString object matching JS: { payload: requestBody, moduledata }
        let payloadValue: Any = requestBody ?? ""
        let moduleDataMap: [String: Any] = [
            "payload": payloadValue,
            "moduledata": moduledata
        ]

        // JSON.stringify equivalent
        let stringToHash: String
        if let jsonData = try? JSONSerialization.data(withJSONObject: moduleDataMap, options: [.sortedKeys]),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            stringToHash = jsonString
        } else {
            stringToHash = "{}"
        }

        // Append salt (IV string)
        let combinedString = stringToHash + salt

        // SHA-256 hash and hex-encode
        return sha256Hex(combinedString)
    }

    private static func sha256Hex(_ input: String) -> String {
        let data = Data(input.utf8)
        let hash = SHA256.hash(data: data)
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}
